### ISSUE TITLE
OVAL ID naming schemes

### ISSUE THREAD

**Issue by wmunyan** [CISecurity/OVALRepo#40]
Question - Was it ever decided how OVAL IDs would be handled using the GitHub model?  I know it was decided that CIS wouldnt mass-change all the "oval.mitre.org" IDs to another namespace.

Were we going to be accepting submissions from contributors using an ID namespace of their own choosing?  This has the benefit of allowing contributors to maintain their own pools of IDs and put the burden of ID maintenance on each contributor.

Or should we require all submissions of NEW content to utilize an "oval.cisecurity.org" namespace?  This would require contributors to re-syncronize their forks and utilize scripts to ensure any NEW content being submitted does NOT have an ID already existing in the repository.

I would really like folks' thoughts on this as it has a definite impact on the submission processing.

**Comment by DavidRies** [comment]
I have no opinion on this. I don't really understand why everything was NS'd oval.mitre.org historically. Maybe ask the MITRE folks if there's a good reason for this that applies to us as well?

**Comment by GunnarEngelbach** [comment]
The Mitre process is that submitters will make up their own IDs in the 
submitted content, but that on acceptance Mitre would replace that with 
an ID using "oval.mitre.org"

That seems like a good compromise.   The repository owner can't 
necessarily have reliance on contributors to be managing their own IDs 
well enough to be able to guarantee uniqueness once they are in the full 
repository.

**Comment by wmunyan** [comment]
For me, this is an issue in submission processing.  A set of decomposed files is contained in a PR.  That PR can be merged or rejected.  If the PR is merged, then I am not aware of any method to be able to modify the contents of the PR before it gets merged.  Does that make sense?  If the submission processing is going to change IDs, then is there a way to make those ID changes before the PR is merged into the repository.  If the contents of the PR cannot be changed within the PR, then that PR cannot be merged; it will have to be sucked down to a CIS local repository, IDs massaged, and then committed/pushed by CIS.

Or is there a way to close the PR in such a way that contributors wont continue to see prior submissions in subsequent PRs?

**Comment by GunnarEngelbach** [comment]
The choices are to trust contributors to properly manage OVAL IDs, or to 
be able to edit submissions prior to final acceptance.

Since it will also be necessary to update other things in submitted 
content (<status> elements, for example, and version numbers) we might 
as well edit OVALIDs as well.

Yes, this means that a pull request just can't be pushed into the 
upstream master on acceptance.  Some level of editing will need to be 
done first.

**Comment by DavidRies** [comment]
I agree. A few things will probably need to happen before PR is accepted and published: QA, ID massage?, potentially updating oval_repository metadata like status and contributor, etc.

I assumed these steps would happen as follows:
1. CIS merges PR into a local working copy of the repo via CLI (not publicly available on github)
2. CIS performs QA, massage, etc. on that local working copy
3. If there are QA issues, CIS notes in PR and updates local working copy when PR is updated (back to 2)
4. Once everything is QA'd and massaged, CIS pushes local merged to public repo

Does that make sense and work for you guys or do we need a different process?

**Comment by wmunyan** [comment]
David, 
That makes sense and is a general outline of what my submission processing would be.  Im glad Im on the same page; makes me feel like I'm starting to understand something.

So then what happens to the original PR?  Does it just get closed?  What downstream effects does simply closing a PR and not merging the original PR have on those who have submitted the PR?  I think I understand that if they use a "feature branch" that they could just delete that branch and be no worse off?  I dont want a closed-but-not-merged PR to always be sitting in their next PR.  Does that make sense?

**Comment by DavidRies** [comment]
Cool. I think that when you push the merged PR to the public repo, it will close the PR (i.e. have the same effect as using GitHub UI to accept/merge a PR). That should notify everyone. Then, as described in step 10 of https://github.com/CISecurity/OVALRepo/blob/master/CONTRIBUTING.md : 

"Clean Up: once your PR has been accepted, you should delete your topic branch and update your master branch from the upstream/master so you have all the latest content (including your submission)!"

**Comment by wmunyan** [comment]
I think the difference in my thinking is that, because changes to the content would always need to be made, such as IDs, versions, statuses, etc, there would NEVER be a "merged PR", because if the original PR is merged, it wouldnt contain the massaged IDs, versions, statuses, etc.  The PR would just have to be closed without merging, and would there then be downstream effects for contributors because of that...

**Comment by DavidRies** [comment]
So, pending actually trying this :), here's how I think it works:
1. you do an actual merge into your local copy and that's committed locally
2. you do any changes for IDs etc and QA until it's ready for publication and commit all that
3. then you, when you push, you're pushing the merge and the published form (it will all be in commit history) so there is a real merge that get's pushed live and I think that closes the PR etc.

**Comment by solind** [comment]
Doesn't changing IDs mean changing file names, which eliminates the 'gitfulness' of the whole process?

**Comment by GunnarEngelbach** [comment]
Changing the ID would only happen for a new element.

**Comment by solind** [comment]
Ok then, we should document a best practice for creating IDs for new elements (i.e., start with your company's reverse DNS name), and CIS can correct any issues -- or ask the submitter to make changes which will update the PR automatically.

I don't necessarily want a process where CIS has to do all the work, unless they really want to do it.

**Comment by wmunyan** [comment]
David S - Yes, I think changing the IDs eliminates the "gitfulness" of the process, that's the purpose behind the question I asked originally.  Are we expecting contributors to maintain their own ID pools based on their namespace (reverse DNS).

I would certainly like contributors to maintain their own ID pools.  Is that something we can require from contributors as part of the submission instructions?

**Comment by GunnarEngelbach** [comment]
On 7/16/2015 8:26 PM, Bill M wrote:
>
> David S - Yes, I think changing the IDs eliminates the "gitfulness" of 
> the process, that's the purpose behind the question I asked 
> originally. Are we expecting contributors to maintain their own ID 
> pools based on their namespace (reverse DNS).
>
> I would certainly like contributors to maintain their own ID pools. Is 
> that something we can require from contributors as part of the 
> submission instructions?
>

Certainly you can, and to some degree they are already doing that 
because they use their own IDs when they submit to Mitre.

If they are using the git fork-n-pullrequest submission method they will 
have to manage their own IDs.  Otherwise they would overwrite their own 
content.

However, if you want to use those IDs as the final IDs in the master 
repository, you'll still need to verify that they are unique and be able 
to change them when they aren't.  While you can require that submitters 
manage their IDs, you can't guarantee that with multiple submitters from 
the same company that they will be 100% reliable at doing that.

**Comment by solind** [comment]
I have to say, I don’t really care what IDs are used (provided there are no collisions).  It would be nice if a contributor decided to use their own virtual namespace and for the most part managed it themselves, but it’s not at all necessary for the repository to work.

I think, unless someone overwrites an existing object and makes something completely new out of it (as opposed to fixing or refining it), we shouldn’t really care what ID they use.  OTOH, I shouldn’t be creating com.threatguard IDs and Gunnar shouldn’t be creating org.joval IDs.  And Individual contributors who aren’t tied to an organization will need a namespace too.  They shouldn’t necessarily use, for example, com.yahoo or com.gmail because their personal email account is hosted there.

So my conclusion is, we should have some equivalent to how org.mitre was used — as a catch-all for people who for whatever reason didn’t choose to use their own organization.  And we should also allow contributors to also use an ID name-space that’s related to their organization, if they have one.

This is all just for contributing brand-new content.  Edits to existing content should take the form of fixes, corrections and enhancements only, existing IDs should be maintained, and versions rev’d.

**Comment by DavidRies** [comment]
decided: we will use CIS NS for updates.
